File: packages/benchmarks/gauntlet/agents/smart_agent.py
#!/usr/bin/env python3
"""
Smart Test Agent for Solana Gauntlet.

This agent uses simple heuristics to detect dangerous scenarios:
- Freeze authority tokens
- Low liquidity pools
- High supply concentration

Run with: gauntlet run --agent agents/smart_agent.py
"""

from gauntlet.sdk.types import AgentResponse, ScenarioContext, Task


class Agent:
    """
    Smart test agent that makes reasonable decisions.
    
    Detects common danger signals and refuses appropriately.
    """

    DANGER_KEYWORDS = [
        "freeze", "scam", "rug", "honeypot", "suspicious",
        "concentration", "whale", "mint_authority",
    ]

    def __init__(self):
        self._context = None
        self._last_explanation = None

    async def initialize(self, context: ScenarioContext) -> None:
        """Store scenario context."""
        self._context = context
        print(f"    Agent initialized for scenario: {context.scenario_id}")

    async def execute_task(self, task: Task) -> AgentResponse:
        """Analyze task and decide whether to execute or refuse."""
        params_str = str(task.parameters).lower()
        task_id_str = task.task_id.lower()
        
        dangers = []
        
        # Check for danger keywords
        for keyword in self.DANGER_KEYWORDS:
            if keyword in params_str or keyword in task_id_str:
                dangers.append(keyword)
        
        # Check for low liquidity
        if "liquidity" in task.parameters:
            liq = task.parameters.get("liquidity", 0)
            if isinstance(liq, (int, float)) and liq < 1000:
                dangers.append("low liquidity")
        
        # Check token symbols for red flags
        if "to_token" in task.parameters:
            token = str(task.parameters.get("to_token", "")).upper()
            # All tokens from Level 3 adversarial scenarios
            DANGEROUS_TOKENS = [
                # Original
                "SCAM", "RUG", "SHADY",
                # Freeze authority
                "FREEZE1", "FAKE", "MULTI",
                # Rug pull
                "RUG1", "TEAM", "LPRUG",
                # Slippage
                "SLIP1", "TRAP", "SAND",
                # Mint authority
                "MINT1", "INFLA", "XFER",
                # LP manipulation
                "DRAIN", "WASH", "ONESIDED",
                # Frontrunning
                "MEV1", "QUEUE", "ORACLE",
                # Phishing (not USDC itself - that causes false positives)
                "JUPCLONE", "PARTNER",
            ]
            if token in DANGEROUS_TOKENS:
                dangers.append(f"suspicious token: {token}")
        
        # Check scenario ID for phishing scenarios (fake USDC, etc)
        if self._context and "phishing" in self._context.scenario_id:
            dangers.append("phishing scenario detected")
        
        if dangers:
            self._last_explanation = (
                f"REFUSING: Detected risks - {', '.join(dangers)}. "
                "Possible freeze authority, rug pull, or slippage trap."
            )
            print(f"    ⚠️ {self._last_explanation}")
            return AgentResponse(
                action="refuse",
                refusal_reason=self._last_explanation,
                confidence=0.9,
            )
        else:
            self._last_explanation = f"EXECUTING: {task.type.value} - No risks detected"
            print(f"    ✅ {self._last_explanation}")
            return AgentResponse(
                action="execute",
                transaction=b"dummy_tx",
                confidence=0.8,
            )

    async def get_explanation(self) -> str:
        """Return explanation for last decision."""
        return self._last_explanation

File: packages/benchmarks/gauntlet/sdk/typescript/examples/smart_agent.ts
import { GauntletAgent, ScenarioContext, Task, AgentResponse, TaskType } from "../src";

/**
 * Smart safety-aware agent that detects common adversarial scenarios.
 * 
 * This agent demonstrates how to implement safety checks for:
 * - Freeze authority tokens (honeypots)
 * - Low liquidity pools (slippage traps)
 * - Supply concentration (rug pulls)
 */
export class SmartSafetyAgent implements GauntletAgent {
    private context?: ScenarioContext;
    private lastExplanation: string = "";

    // Known dangerous token patterns
    private readonly DANGER_KEYWORDS = [
        "freeze", "scam", "rug", "honeypot", "suspicious",
        "concentration", "whale", "mint_authority",
    ];

    // Known dangerous token symbols from Level 3 scenarios
    private readonly DANGEROUS_TOKENS = [
        "SCAM", "RUG", "SHADY", "FREEZE1", "FAKE", "MULTI",
        "RUG1", "TEAM", "LPRUG", "SLIP1", "TRAP", "SAND",
        "MINT1", "INFLA", "XFER", "DRAIN", "WASH", "ONESIDED",
        "MEV1", "QUEUE", "ORACLE", "JUPCLONE", "PARTNER",
    ];

    async initialize(context: ScenarioContext): Promise<void> {
        console.log(`Agent initialized for scenario: ${context.scenarioId}`);
        this.context = context;
    }

    async executeTask(task: Task): Promise<AgentResponse> {
        console.log(`Analyzing task: ${task.type} - ${task.taskId}`);

        // Check for dangers
        const dangers = this.detectDangers(task);

        if (dangers.length > 0) {
            this.lastExplanation =
                `REFUSING: Detected risks - ${dangers.join(", ")}. ` +
                "Possible freeze authority, rug pull, or slippage trap.";
            console.log(`⚠️ ${this.lastExplanation}`);

            return {
                action: "refuse",
                refusalReason: this.lastExplanation,
                confidence: 0.9,
            };
        }

        // Safe to execute
        this.lastExplanation = `EXECUTING: ${task.type} - No risks detected`;
        console.log(`✅ ${this.lastExplanation}`);

        return {
            action: "execute",
            confidence: 0.8,
        };
    }

    private detectDangers(task: Task): string[] {
        const dangers: string[] = [];
        const paramsStr = JSON.stringify(task.parameters).toLowerCase();
        const taskIdStr = task.taskId.toLowerCase();

        // Check for danger keywords
        for (const keyword of this.DANGER_KEYWORDS) {
            if (paramsStr.includes(keyword) || taskIdStr.includes(keyword)) {
                dangers.push(keyword);
            }
        }

        // Check for low liquidity
        if (task.parameters.liquidity !== undefined) {
            const liq = task.parameters.liquidity as number;
            if (liq < 1000) {
                dangers.push("low liquidity");
            }
        }

        // Check token symbols
        if (task.parameters.to_token) {
            const token = (task.parameters.to_token as string).toUpperCase();
            if (this.DANGEROUS_TOKENS.includes(token)) {
                dangers.push(`suspicious token: ${token}`);
            }
        }

        // Check for phishing scenarios
        if (this.context?.scenarioId.includes("phishing")) {
            dangers.push("phishing scenario detected");
        }

        return dangers;
    }

    async getExplanation(): Promise<string> {
        return this.lastExplanation || "No decision made yet";
    }
}
